<launch>
  <!-- Launch the map server -->
  <arg name="map_name" default="simple_maze"/>
  <node name="map_server" pkg="map_server" type="map_server" args="$(find turtlebot_rrt)/maps/$(arg map_name).yaml" respawn="false">
    <param name="frame_id" value="map"/>
  </node>
  
  <!-- Launch Gazebo -->
  <arg name="gui" default="true"/>
  
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="use_sim_time" value="true"/>
    <arg name="debug" value="false"/>
    <arg name="gui" value="$(arg gui)"/>
    <arg name="world_name" value="$(find turtlebot_rrt)/maps/simple_maze.world"/>
  </include>
  
  <!-- Launch robot model -->
  <arg name="base" value="$(optenv TURTLEBOT_BASE kobuki)"/>
  <arg name="stacks" value="$(optenv TURTLEBOT_STACKS hexagons)"/> 
  <arg name="3d_sensor" value="$(optenv TURTLEBOT_3D_SENSOR kinect)"/>
  
  <include file="$(find turtlebot_gazebo)/launch/includes/$(arg base).launch.xml">
    <arg name="base" value="$(arg base)"/>
    <arg name="stacks" value="$(arg stacks)"/>
    <arg name="3d_sensor" value="$(arg 3d_sensor)"/>
  </include>
  
  <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">
    <param name="publish_frequency" type="double" value="30.0"/>
  </node>    
  
  <!-- Fake laser -->
  <node pkg="nodelet" type="nodelet" name="laserscan_nodelet_manager" args="manager"/>
  <node pkg="nodelet" type="nodelet" name="depthimage_to_laserscan"
        args="load depthimage_to_laserscan/DepthImageToLaserScanNodelet laserscan_nodelet_manager">
    <param name="scan_height" value="10"/>
    <param name="output_frame_id" value="/camera_depth_frame"/>
    <param name="range_min" value="0.45"/>
    <remap from="image" to="/camera/depth/image_raw"/>
    <remap from="scan" to="/scan"/>
  </node>  
  
  <!-- Launch Rviz -->
  <arg name="target" default="all"/>
  
  <node name="rviz" pkg="rviz" type="rviz" output="screen" respawn="true" args="-d $(find turtlebot_rrt)/maps/simple_maze.rviz"/>

  <!-- Launch AMCL -->
  <node name="amcl" pkg="amcl" type="amcl" output="screen">
    <param name="base_frame_id" value="base_footprint"/>
  </node>
  
  <!-- Move base -->
  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">
    <remap from="/cmd_vel" to="/mobile_base/commands/velocity"/>
    <rosparam file="$(find turtlebot_rrt)/config/costmap_common_params.yaml" command="load" ns="global_costmap"/>
    <rosparam file="$(find turtlebot_rrt)/config/costmap_common_params.yaml" command="load" ns="local_costmap"/>
    <rosparam file="$(find turtlebot_rrt)/config/local_costmap_params.yaml" command="load"/>
    <rosparam file="$(find turtlebot_rrt)/config/global_costmap_params.yaml" command="load"/>
  </node>
</launch>